File: src/utils/logger.js
const level = (process.env.LOG_LEVEL || 'info').toLowerCase();
const LEVELS = { debug: 0, info: 1, warn: 2, error: 3 };

function log(lvl, msg, meta = {}) {
  if (LEVELS[lvl] < LEVELS[level]) return;
  const line = {
    ts: new Date().toISOString(),
    level: lvl,
    msg,
    ...meta,
  };
  console.log(JSON.stringify(line));
}

export const logger = {
  debug: (msg, meta) => log('debug', msg, meta),
  info:  (msg, meta) => log('info',  msg, meta),
  warn:  (msg, meta) => log('warn',  msg, meta),
  error: (msg, meta) => log('error', msg, meta),
};

File: backend/logger.py
import json
import sys
from datetime import datetime, timezone
from config import CONFIG

LEVELS = {"debug": 0, "info": 1, "warn": 2, "error": 3}
_current_level = LEVELS.get(CONFIG.log_level.lower(), 1)


def _log(level: str, msg: str, meta: dict | None = None) -> None:
    if LEVELS.get(level, 0) < _current_level:
        return
    line = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "level": level,
        "msg": msg,
        **(meta or {}),
    }
    print(json.dumps(line), file=sys.stdout, flush=True)


class Logger:
    def debug(self, msg: str, meta: dict | None = None) -> None:
        _log("debug", msg, meta)

    def info(self, msg: str, meta: dict | None = None) -> None:
        _log("info", msg, meta)

    def warn(self, msg: str, meta: dict | None = None) -> None:
        _log("warn", msg, meta)

    def error(self, msg: str, meta: dict | None = None) -> None:
        _log("error", msg, meta)


logger = Logger()
